DM-22508: add Database class, concrete implementations, and tests

TallJimbo opened the pull request:
The code here is not used at all in Butler yet; it's the beginning of a new backend for Registry, and hence there's some temporary code duplication with existing Registry code.

The testing strategies for PostgreSQL and Oracle merit some special attention:

- For PostgreSQL, we use the `testing.postgresql` module (available via `pip`)  to spin up a local userspace database server to test against.  We do that in `setUpClass` to avoid doing it repeatedly, and then create one or more schemas within that database _per test function_.  We delegate cleaning that all up in `tearDownClass` to `testing.postgresql`.  Tests are skipped if `testing.postgresql` cannot be imported, and we assume that the presence of `testing.postgresql` indicates that the actual PostgreSQL server binaries are in the PATH, because `testing.postgresql` can't ever do anything unless that's true, even though it's not taken care of by `pip` (the `postgresql` conda package does take care of it, however).

- For Oracle, we look for a `DAF_BUTLER_ORACLE_TEST_URI` environment variable that provides the connection string to pass to SQLAlchemy.  If that's set, we make a single connection to the database server (in `setUpClass`) and use (as @MichelleGower has advocated) a randomly-generated prefix for all tables we create in that schema to simulate a virtual schema.  There's custom code in `tearDownClass` that then drops all tables that were created, using a Python DBAPI rewrite of the `clobber.sql` script at https://confluence.lsstcorp.org/display/DM/Oracle+with+Gen3.

While this basically works, it's worth pointing out the disadvantages of the Oracle approach; while the PostgreSQL is obviously much harder to pull off with any proprietary database engine, it'd be really nice to set up something more like that for Oracle, even if it's only something we can interact with via some CI server.  In particular:
 - SQLAlchemy basically takes care of real schemas for us, and we sort of have to fight against it to use table prefixes to make those work like a virtual schema.  The need to use prefixes actually adds more complexity to the `OracleDatabase` specialization than anything else unique to Oracle.
 - By not using real schemas in the Oracle tests, we don't actually get test coverage for trying to use real schemas in Oracle, which we of course really want to do in the future.
 - While I'm pretty sure test cleanup usually happens correctly when tests fail, I'd guess there are myriad ways to interrupt the tests such that the cleanup is at best incomplete.  And that of course means one has to go in and clean things up manually.
 - Running tests against a local server is lot faster - or at least I assume that's the reason for the pretty huge gap between the ~1.6s PostgreSQL tests and Oracle tests, which I've seen take anything between 20s and 3 minutes (it might also be something silly I've done, or overhead in Oracle in creating tables - which tests naturally do a lot more of than production use would).

TallJimbo committed:
Copy stripped-down version of schema dataclasses to new subpackage.

The new 'registry' subpackage will be the home for classes that are
entirely internal to the new Registry.  Duplicate code in the 'core'
package will eventually be removed as the old Registry is replaced.

TallJimbo committed:
Add Database ABC with table definition API only (so far).

TallJimbo committed:
Add SQLite Database implementation and tests.

TallJimbo committed:
Add utility class to truncate too-long names in Databases.

TallJimbo committed:
Add PostgreSQL Database implementation.

TallJimbo committed:
Add insert and query methods to Database.

TallJimbo committed:
Add delete method to Database.

TallJimbo committed:
Add update method to Database.

TallJimbo committed:
Ensure SQLite autoincrement columns are exactly Integer.

TallJimbo committed:
Add workaround for compound primary keys with autoincrement in SQlite.

TallJimbo committed:
Add unit tests for Database.insert, query, and delete.

TallJimbo committed:
Add Oracle implementation for Database.

TallJimbo committed:
Add name shrinking to PostgresqlDatabase.

DinoBektesevic commented on python/lsst/daf/butler/registry/databases/sqlite.py:
```diff
@@ -0,0 +1,217 @@
+def _replace(insert, compiler, **kw):
+    """Generate an INSERT ... ON CONFLICT REPLACE query.
+
+    Unlike the PostgreSQL case, sqlalchemy does not provide special support for
```
Is "Unlike the PostgreSQL case" neccessary?

TallJimbo replied:
I think it helps explain why this code is here (it and the code is copied from something @andy-slac wrote earlier), but maybe it's better as a code comment than a docstring.

DinoBektesevic commented on python/lsst/daf/butler/registry/databases/sqlite.py:
```diff
@@ -0,0 +1,217 @@
+
+
+@dataclass
+class _AutoincrementWorkaround:
```
Rename to "_AutoincrementCompoundKey"? I understand this is, partially, internal and is used only once but would it still make code that uses it easier to read?

TallJimbo replied:
I'll go with `_AutoincrementCompoundKeyWorkaround`, because I don't want to imply that it *is* the compound key.

DinoBektesevic commented on python/lsst/daf/butler/registry/databases/sqlite.py:
```diff
@@ -0,0 +1,217 @@
+            )
+        self._connection.execute(_Replace(table), *rows)
+
+    filename: Optional[str]
```
I'm confused, if this is a class attribute shouldn't it be on top under class docstring?

TallJimbo replied:
(see previous comment thread on this subject)

DinoBektesevic commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+
+def _checkExistingTableDefinition(name: str, spec: ddl.TableSpec, inspection: Dict[str, Any]):
+    """Test that the definition of a table in a `ddl.TableSpec` and from
+    database introspection are the consistent.
```
"the" is extra?

andy-slac replied:
extraneous `the`.

DinoBektesevic commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        # TODO: should we guard against non-SELECT queries here?
+        return self._connection.execute(sql, *args, **kwds)
+
+    origin: int
```
I went looking through the DM Style-guide and couldn't find a place that mentions positioning of class attributes so I assume this is ok too. I am confused though, is this a new thing everyone should be doing? I thought they should all go to the top?

TallJimbo replied:
:man_shrugging: I've been putting them here just because I've been sold on the idea that it makes sense for `__init__` to go at the top so it immediately follows the class docstring, and then I like keeping other methods with it.  I had assumed there was no convention or strong precedent one way or the other (especially for these, which are formally class attributes but really exist just as a place to hang instance attribute documentation).

DinoBektesevic replied:
I was asking just in case I make a PR again.... Thanks.

DinoBektesevic commented on tests/test_postgresql.py:
```diff
@@ -0,0 +1,130 @@
+
+    def makeEmptyDatabase(self, origin: int = 0) -> PostgresqlDatabase:
+        namespace = f"namespace_{self.testCounter}_{self.dbCounter}"
+        self.dbCounter += 1
```
Pytest parallel tests run in separate processes, so the counter won't be shared but it won't be unique either. It was my impression from reading the code that namespace needs to be separate to avoid conflicts because this is all executing in the same DB. 

Did I miss-understand? If not, should this be replaced by something that gives unique counters?

TallJimbo replied:
I'm pretty sure pytest would also have to run `setUpClass` in each process or `cls.server` (etc) won't exist sometimes and the tests wouldn't run at all (unless it's pickling the `TestCase`, I guess - I hope it isn't).  In any case, it's easy to change to something random here, and I'll just do that anyway.

andy-slac commented on python/lsst/daf/butler/registry/tests/database.py:
```diff
@@ -0,0 +1,387 @@
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
```
Probably needs `__all__` here?

andy-slac commented on python/lsst/daf/butler/registry/nameShrinker.py:
```diff
@@ -0,0 +1,70 @@
+class NameShrinker:
+    """A utility class for `Database` implementations that need a nontrivial
+    implementation of `Database.shrinkDatabaseEntityName` and
+    `Database.expandDatabaseEntityName`.
```
If all users of this class are implementation classes in `database` package maybe this module should be there too?

TallJimbo replied:
My intent was to make all of the files in the `databases` package single-class modules that define a concrete `Database` subclass.  Sounds like maybe I should add a README or something to that directory so that's not just in my head, at least.

andy-slac commented on python/lsst/daf/butler/registry/ddl.py:
```diff
@@ -0,0 +1,191 @@
+from math import ceil
+from dataclasses import dataclass
+from typing import Optional, Tuple, Set
+import typing
```
Looks like this import is used for `typing.Sequence`, I'm not sure why `Sequence` is treated differently from `Tuple` or `Set` imported above, can you do all of them in the same style?

TallJimbo replied:
Will do.  It's historical; over in the prototype version (where this is still in `daf/butler/core/schema.py`) we also do `from sqlalchemy.schema import Sequence`, but I figured just importing sqlalchemy without lifting its symbols was usually more readable here.

andy-slac commented on python/lsst/daf/butler/registry/ddl.py:
```diff
@@ -0,0 +1,191 @@
+            return None
+        if not isinstance(value, bytes):
+            raise TypeError(
+                f"Base64Bytes fields require 'bytes' values; got {value} with type {type(value)}"
```
Put quotes around `{value}` in case it's empty or very long?

andy-slac commented on python/lsst/daf/butler/registry/ddl.py:
```diff
@@ -0,0 +1,191 @@
+@dataclass
+class TableSpec:
+    """A struct-like class used to define a logical `Registry` table (which may
+    be implemented in the database as a query on any number of tables).
```
Not sure how to read this right, does "query" mean a regular SQL view here or does it mean that we can represent schema of the regular query with this structure?
If it can be used for schema of arbitrary query then one potential issue is that column names in the query result may not be unique (this obviously depends on particular backend and a query) and using NamedValueSet for fields can be an issue. I do not know if this is an issue for registry schema, just a general comment.

TallJimbo replied:
The idea was that this class is often used to communicate to `Database` and other classes what columns their users expect to be able to query.  The `Database` (or whatever is given the `TableSpec`) is then free to implement that query interface via a true table, a view, or even just Python/SQLAlchemy object that returns a query that could be used as a subquery (essentially a view implemented in Python instead in the database).

Since that's really about how this class is used by other classes, maybe I should just remove most of that discussion from these docs.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+                           f"table in the database has {columnNames}.")
+
+
+class TransactionInterruption(RuntimeError):
```
I think exception classes should end with "Error". Same for `SynchronizationConflict` below.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+    if spec.fields.names != set(columnNames):
+        raise RuntimeError(f"Table '{name}' exists but is defined differently in the database; "
+                           f"specification has columns {list(spec.fields.names)}, while the "
+                           f"table in the database has {columnNames}.")
```
Do you plan checking for types in the future?

TallJimbo replied:
No plans right now.  One issue I know of is that SQLite actually requires us to change `BigInteger` and `SmallInteger` to just `Integer` in order for autoincrement to work, but then it doesn't actually care how big the numbers are that we put in those fields, so checking for exact types there would not work.  Even for databases with more strict typing, many database types will map to the same Python type, so checking at this point naively might break some otherwise-backwards-compatible migrations.  And the real goal here is to catch cases where the Python code is out of sync with what's in the database, and I think that's unlikely to just involve just type changes.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+            )
+        return table
+
+    def addTableTuple(self, specs: Tuple[ddl.TableSpec]) -> Tuple[sqlalchemy.schema.Table]:
```
Should it be `NamedTuple` instead of `Tuple`?

TallJimbo replied:
Ideally, yes, this would be `NamedTuple[ddl.TableSpec, ...]` (I just realized I forgot the `...` above).  But `typing.NamedTuple` apparently wasn't intended to be used for this ([the docs](https://docs.python.org/3/library/typing.html#typing.NamedTuple) only show usage in defining a new `NamedTuple` class), and it doesn't work in mypy.  I guess that's because `collections.namedtuple` is just a type factory, not a base class or even a metaclass.

I'll add a note about this to the docs.

andy-slac reviewed (commented):
I only started looking into it, but cannot finish today. Here are the comments that I have so far, will continue tomorrow.

DinoBektesevic reviewed (commented):
I don't see anything obviously wrong but I have only started walking through the tests this afternoon. I've noticed Andy pushing his review and I've noticed we overlap in places I'm adding questions I had when I read through it. 

Thanks for simultaneously fixing the name length limitations for PostgreSQL, that helps me a lot!

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+            transaction to be rolled back or committed independent of any
+            outer transaction.  If `False` (default), transactions may still
+            be nested, but nothing will be committed until the outermost
+            transaction context ends.
```
It's phrased a bit confusingly, with SAVEPONT commit does not actually happen until outermost (regular) transaction is committed. Main difference between those is in rollback, with SAVEPOINT rollback means "rollback to savepoint", while with nested regular transaction rollback aborts outermost transaction.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        """
+        if interrupting and self._transactions:
+            raise TransactionInterruption("Logic error in transaction nesting: an operation that would "
+                                          "the active transation context has been requested.")
```
I think there is a missing word in this message.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        instances.
+    connection : `sqlalchemy.engine.Connection`, optional
+        SQLAlchemy connection object.  May be shared with other `Database`
+        instances.  Constructed from ``engine`` if not provided.
```
Do we really need connections shared between Databases? With shared connection some logic in this class may be broken, e.g. `_transactions` may be empty but transaction may actually be in progress if started by another Database instance.

TallJimbo replied:
The idea was that these would be different `Database` instances in the same multi-layer `Registry`, each pointing to a different schema in the same server.  I think there sharing a connection is desirable.  It's definitely not okay for different `Butler` instances to share a connection.

andy-slac replied:
Still, if the same connection instance is shared between two databases then there is a potential that state of one Database instance (`_transactions` list) may not reflect actual state of the connection which was modified by other Database instance. Maybe instead of tracking transactions in `_transactions` list you could just check whether connection is already in transaction?

TallJimbo replied:
Hmm, that's a good point.  I still think that this probably would be easy to avoid in the future `Registry` code that I expect to be the only place shared connections are used (other than unit tests), but this does feel like a problem in the interface if we have to rely on code outside of `Database` behaving a certain way in order for its behavior to be correct.  I'll see if I can guarantee that the `_transactions` list is shared between `Databases` if and only if the `_connection` is.

TallJimbo replied:
I've pushed one more big commit, which adds some factory `classmethods` and a new `dataclass` that ensures the engine, connection and list of transactions is passed together.  It's probably still possible to misuse, but I think it's now a lot harder to *accidentally* misuse, and the checks for whether a transaction is active should now be safe.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+
+        Parameters
+        ----------
+        create : `bool` If `True`, attempt to create all tables at the end of
```
Need a line break after type, same in Returns section

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+            for table, foreignKey in context._foreignKeys:
+                table.append_constraint(foreignKey)
+            if create:
+                if create and self.namespace is not None:
```
`create` is tested on the line above already

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+                        self._engine.execute(sqlalchemy.schema.CreateSchema(self.namespace))
+                self._metadata.create_all(self._engine)
+        except BaseException:
+            self._metadata.drop_all(self._engine)
```
This looks dangerous, should we only drop tables that were created by previous call to `create_all` (not sure if we can find which ones existed before this call)?

TallJimbo replied:
I think this already does what we want - it only drops tables known to the `MetaData` object, and that can only be the tables we've created since the start of the `with db.declareStaticTables` context, since we create the `MetaData` object at the start of that.

andy-slac replied:
Hmm, I though that metadata `create_all` just ignores re-existing tables. So if there were some table in the database and in metadata, `drop_all` would remove those even if they existed before. I guess this is probably not an issue as we either should have all tables in a database or none.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+
+    def shrinkDatabaseEntityName(self, original: str) -> str:
+        """Return a version of the given name that fits within this database
+        engine's length limits for table, constraint, indexe, and sequence
```
`indexe`

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        metadata : `sqlalchemy.MetaData`
+            SQLAlchemy representation of the DDL schema this field's table is
+            being added to.
+        kwds
```
I have no strong opinion on this but our doc examples show this should be `**kwds`: https://developer.lsst.io/python/numpydoc.html#parameters

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        Parameters
+        ----------
+        spec : `TableSpec`
+            Specification for the foreign key to be added.
```
foreign key?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        ReadOnlyDatabaseError
+            Raised if `isWriteable` returns `False`, and the table does not
+            already exist.
+        RuntimeError
```
Given that TransactionInterruption and ReadOnlyDatabaseError inherit from RuntimeError there is a potential for confusion if someone write code like this:

    try:
        db.ensureTableExists(...)
    catch RuntimeError:
        log.error("Where did you get this spec, it's inconsistent with database schema!")
    catch TransactionInterruption
        log.error("What are you doing, database is already in transaction!")

TallJimbo replied:
I [pinged slack](https://lsstc.slack.com/archives/C2JPL8U78/p1576168318008800) about what kinds of exceptions to make these.  I think I'm going to replace `TransactionInterruption` with just using `assert`, and I think I might generalize `SynchronizationConflict` to `DatabaseConflictError`, since from the user perspective inconsistent row data and inconsistent DDL for a dynamic table aren't meaningfully different.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+                    if n < 1:
+                        raise RuntimeError("Insertion in sync did not seem to affect table.  This is a bug.")
+                    elif n > 2:
+                        raise RuntimeError("Keys passed to sync do not comprise a unique constraint.")
```
Maybe give a list of key columns and table name to simplify debugging?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+                # Most likely cause is that an equivalent row already exists,
+                # but it could also be some other constraint.  Query for the
+                # row we think we matched to resolve that question.
+                n, bad, result = check()
```
Do you need `self.transaction()` around this line?

TallJimbo replied:
I didn't think so, but you might know better than I do - `check` only executes a `SELECT` query, and I didn't think there was any point to putting that in a transaction (except maybe in SQLite when using super-aggressive locking, but I don't want to rely on that going forward, and I don't think we need to).

andy-slac replied:
Well, this code should execute outside any transaction (code above uses interrupting=True), and if you don't manage transactions yourself then it is probably sqlalchemy auto-commit feature that handles transactions for you (https://docs.sqlalchemy.org/en/13/core/connections.html#understanding-autocommit). I think it's better to manage transactions explicitly, especially if you tracking active transactions in your class.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        if not returnIds:
+            self._connection.execute(table.insert(), *rows)
+        else:
+            return [self._connection.execute(table.insert(), row).inserted_primary_key[0] for row in rows]
```
Maybe move `table.insert()` out of the loop for small optimization.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        ----------
+        table : `sqlalchemy.schema.Table`
+            Table rows should be inserted into.
+        rows
```
`*rows` maybe?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        ----------
+        table : `sqlalchemy.schema.Table`
+            Table that rows should be deleted from.
+        keys: `~collections.abc.Iterable` of `str`
```
keys -> columns

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+            be deleted; these will be combined via ``AND`` to form the
+            ``WHERE`` clause of the delete query.
+        rows
+            Positional arguments are the keys of rows to be deleted, asIterable
```
I think this needs some rephrasing.

TallJimbo replied:
Line shouldn't have even been affected by this commit; must have hit some bad keystrokes in my editor.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+            ``WHERE`` clause of the delete query.
+        rows
+            Positional arguments are the keys of rows to be deleted, asIterable
+        Return
```
Needs an empty line before it.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        Raises
+        ------
+        ReadOnlyDatabaseError
+            Raised if `isWriteable` returns `False` when this method is called.
```
Double backticks for `isWriteable`

TallJimbo replied:
I think it works fine as is; it's a method on the same class, not a parameter.  I will check later when `Database` is used by `Registry` and gets added to the Sphinx build (I don't think that makes sense quite yet).

andy-slac replied:
Sorry, my confusion, I somehow thought `isWriteable` is a parameter to a method, please ignore.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        rows
+            Positional arguments are the rows to be updated, as dictionaries
+            mapping column name to value.  The keys in all dictionaries must
+            be the combination of the names in ``keys`` and ``values``.
```
There are no `values` parameter, I think you may need it.

TallJimbo replied:
As you probably guessed, I used to have a `values` parameter.  I removed it because I didn't need it, and couldn't even easily use it - SQLAlchemy automatically creates implicit `bindparams` with the names of all of the columns in the `VALUES` clause for which keys exist in the dictionaries you pass later.  I wish it didn't because then I could be confident that logic was only happening once, instead of once per row, but there doesn't seem to be a way around it.

Anyhow, given that, I think the right fix here is to change the docs to say,
> The keys in the dictionaries are either the names of columns to be updated or values in the ``keys`` argument.

(and maybe I should rename the `keys` parameter).

andy-slac replied:
I did not know that implicit bindparam, thanks for explainig!

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,927 @@
+        sql = table.update().where(
+            sqlalchemy.sql.and_(*[table.columns[k] == sqlalchemy.sql.bindparam(v) for k, v in keys.items()])
+        )
+        return self._connection.execute(sql, *rows).rowcount
```
I think something is missing here, UPDATE needs `.values()` method too in addition to `.where()`.

TallJimbo replied:
I think `update().values()` (with no args) is equivalent to just `update()`.  Same with `insert()` and `insert().values()`.

andy-slac reviewed (commented):
Few more comments for `Database` class, will continue working on other modules.

andy-slac commented on python/lsst/daf/butler/registry/tests/database.py:
```diff
@@ -0,0 +1,387 @@
+        newDatabase = self.makeEmptyDatabase()
+        with newDatabase.declareStaticTables(create=True) as context:
+            context.addTableTuple(STATIC_TABLE_SPECS)
+        # Try to ensure the dyamic table exists in a read-only version of that
```
`dyamic`

andy-slac commented on python/lsst/daf/butler/registry/tests/database.py:
```diff
@@ -0,0 +1,387 @@
+            tables = context.addTableTuple(STATIC_TABLE_SPECS)
+        self.checkStaticSchema(tables)
+
+        db2 = self.makeEmptyDatabase(origin=2)
```
Should db1 and db2 use different `namespace` parameter rather than `origin`?

TallJimbo replied:
Each call to `makeEmptyDatabase` is guaranteed to use a new (randomly-generated) `namespace` value.  So using a different `origin` here really doesn't do anything, but it does reflect how we would do things in practice.

andy-slac commented on python/lsst/daf/butler/registry/tests/database.py:
```diff
@@ -0,0 +1,387 @@
+        for row in results:
+            self.assertIn(row["name"], ("b1", "b2"))
+            self.assertIsInstance(row["id"], int)
+        self.assertEqual(results[1]["id"], results[0]["id"] + 1)
```
I'm not sure this is guaranteed to be always true, it can break for Oracle RAC for example.

TallJimbo replied:
I'll weaken it to just `self.assertGreater(results[1]["id"], results[0]["id"])`.  Is that safe?

andy-slac replied:
I think it's safer and should work OK with current set of backends that we use.

andy-slac commented on python/lsst/daf/butler/registry/nameShrinker.py:
```diff
@@ -0,0 +1,70 @@
+        If the given name was not passed to `shrink` or was not modified by
+        it, it is returned unmodified.
+        """
+        return self._names.get(shrunk, shrunk)
```
I guess functionality of this method depends on client calling `shrink()` first. Can this requirement be enforced/checked to avoid nasty surprises?

TallJimbo replied:
Only if we also remembered the names of any already-short-enough named that was passed to `shrink`.  This method is really only intended to help with debugging error messages at a pretty low level, and at least right now I don't think it's worth it.

andy-slac commented on python/lsst/daf/butler/registry/databases/sqlite.py:
```diff
@@ -0,0 +1,217 @@
+            if spec.dtype != sqlalchemy.Integer:
+                # SQLite's autoincrement is really limited; it only works if
+                # the column type is exactly "INTEGER".  But it also doesn't
+                # doesn't care about the distinctions between different integer
```
doesn't doesn't

andy-slac commented on python/lsst/daf/butler/registry/databases/sqlite.py:
```diff
@@ -0,0 +1,217 @@
+                for row in rows:
+                    newRow = row.copy()
+                    id = self._connection.execute(autoincr.table.insert()).inserted_primary_key[0]
+                    newRow[autoincr.column] = id
```
What if client provides an explicit value for auto-increment column?

TallJimbo replied:
Glad you caught this.  I was planning to go in and add extra logic for this but forgot (that's why the check for no rows is there).  To handle this properly, we need to place some restrictions on which this workaround applies (which is okay, given what we need):
 - compound primary keys that include an autoincrement value must have exactly one other field in the primary key called "origin"
 - we can only generate autoincrement values when `origin == self.origin`.

There's quite a bit more new code involved in this change now, and I'd appreciate it if you could take another look. It's currently on new commit, but I'll squash it later.

andy-slac replied:
Will look at it later today or tomorrow.

andy-slac replied:
It looks OK to me though I think there is a potential for confusion here. I myself do not understand very well how autoincrement is supposed to work in case of composite key, maybe it's a general problem and not an issue with our implementation. As an example suppose we have a table with `PK=(auto, origin)` and we insert a row `(origin=1)` first which generates `auto=1`. Next we insert explicit `(auto=1, origin=2)` which I think is a valid operation and should work in general, but this will cause issues with our implementation because autoincr table already has `id=1`. I do not know how to solve this, maybe removing a row from autoincr table before insert (or checking that id is not there, which should be cheaper) would work?

TallJimbo replied:
The code on the new commit (59e7189b05239c067e5cdf32465de1dfde9c2867) tries to handle exactly this case, by inserting into the autoincr table only when `origin == self.origin`.  That means your example will work if and only if `self.origin==1`, but we prohibit the caller from inserting e.g. just (`origin=2)` without an explicit `auto` value.  And that limitation is okay for all of the use cases I have identified.

Of course, that's for SQLite, where this is all manually implemented.  I don't know what the behavior is supposed to be for other databases, but PostgreSQL and Oracle pass the tests I have so far, and those (now) cover this case.  My guess is that they will just skip generating some `auto` values that would have been safe, but that's fine.

andy-slac commented on python/lsst/daf/butler/registry/databases/oracle.py:
```diff
@@ -0,0 +1,198 @@
+        SQLAlchemy engine instance to use directly.
+    namespace : `str`, optional
+        Namespace (schema) for all tables used by this database.  May be,
+        `None` which will use the schema implied by ``uri`` or ``enginre``
```
`enginre`

andy-slac commented on tests/test_sqlite.py:
```diff
@@ -0,0 +1,65 @@
+
+    def makeEmptyDatabase(self, origin: int = 0) -> SqliteDatabase:
+        filename = os.path.join(self.root, f"db-{self.counter}.sqlite3")
+        self.counter += 1
```
I agree with Dino that counter-based unique name may be an issue, can you use some random names for that (tempfile.mkstemp)?

andy-slac commented on tests/test_sqlite.py:
```diff
@@ -0,0 +1,65 @@
+from lsst.daf.butler.registry.databases.sqlite import SqliteDatabase
+from lsst.daf.butler.registry.tests import DatabaseTests
+
+TESTDIR = os.path.abspath(os.path.dirname(__file__))
```
I'd probably make this directory name unique too, with some random stuff in it.

TallJimbo replied:
A random part is added below.

andy-slac commented on tests/test_oracle.py:
```diff
@@ -0,0 +1,146 @@
+
+    def makeEmptyDatabase(self, origin: int = 0) -> OracleDatabase:
+        prefix = f"test_{self._token}_{self._testCounter}_{self._dbCounter}_"
+        self._dbCounter += 1
```
Same here, I think there is a chance for collisions depending on how pytest runs. Maybe some random string would work better here?

andy-slac reviewed (approved):
Looks good in general, small bunch of comments, most of them minor. There is a lot of new code and there is a chance I missed something, you can blame me for that later 🙂

TallJimbo committed:
Replace TransactionInterruption with assertions.

These are nonrecoverable logic bugs; an assertion seems appropriate.

TallJimbo committed:
Generalize SynchronizationConflict to DatabaseConflictError.

TallJimbo committed:
Add generic test code for Database interface.

TallJimbo committed:
Add replace method to Database.

TallJimbo committed:
Add sync method to Database.

TallJimbo committed:
Improve Database.insert test coverage.

This now includes another aspect of the compound+autoincrement case
that requires a workaround in SQLite.

TallJimbo committed:
Finish SQLite compound+autoincrement primary key workaround.

Will squash after review.

TallJimbo reviewed (commented):
I think I've addressed all review comments (and rebased).  Lots of replies should be appearing simultaneously.

TallJimbo committed:
Change syntax for documenting *args, **kwargs.

TallJimbo committed:
Rework Database construction patterns.

Transaction state sharing is now much more explicit, and we have the
beginnings of APIs for generically creating Database instances.

DinoBektesevic commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -0,0 +1,1034 @@
+
+    @classmethod
+    @abstractmethod
+    def connect(cls, uri: str, *, writeable: bool = True) -> ConnectionStruct:
```
You guys move too fast, while I'm writing a review 5 new commits are made :). 

It used to be that one would instantiate a Registry from a RegistryConfig. It used to be that it was the job of RegistryConfig's [connectionString](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/core/registryConfig.py#L68) property to resolve the full connection URI from the YAML file, supplementing usernames, ports, password etc. from the `~/.lsst/db-auth.yaml` file as, if at all, necessary based on the db string provided in `butler.yaml`. 

Would it be too much to ask that `connect` optionally accepts a direct URI or an `RegistryConfig` instance? That way one can still depend on the `connectionString` property to resolve, or not, users and passwords. 

I understand, from the [architecture design doc](https://confluence.lsstcorp.org/display/DM/Architectural+Prototype+for+the+New+Gen3+Registry), that `Registry` has to go but it wasn't clear to me if that's the case for `RegistryConfig` also.

TallJimbo replied:
Heh, you're moving too fast, commenting on my commits while I'm still amending them :smiley: .

Anyhow, I was mostly planning to tackle this in a future ticket (when I make `Registry` start to use `Database`), but I am certainly aiming to keep the high-level behavior the same.  I think I'd lean towards leaving these APIs as-is, but use `RegistryConfig` to generate the final connection string to pass to them.  It'd also be responsible for inferring the `Database` subclass (instead of the `Registry` subclass) from what's in the config.  Does that sound like it would work?

DinoBektesevic replied:
That sounds perfect, in fact that sounds to work out exactly like it is right now. The correct registry class is inferred from the `RegistryConfig.getRegistryClass` based on  the mapping provided in `config/registry.yaml`(for future reference).

TallJimbo committed:
Use connection, not engine, to emit and reflect DDL.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -135,6 +137,36 @@ def addTableTuple(self, specs: Tuple[ddl.TableSpec, ...]) -> Tuple[sqlalchemy.sc
+
+    transactions: List[sqlalchemy.engine.Transaction]
+    """A list of active transactions for this connection; an empty list
+    indicates that no transaction is active.
```
I'm looking at how `transactions` is used and the only pattern that I see is checking for non-empty list to decide whether any transaction is in progress. If this is the only thing that you need from this list then I'd probably vote for removing it entirely and using `Connection.in_transaction()` for that test. 
Even with all current fixes I think that logic for handling `transactions` list is not completely water-tight, see comment below.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -208,23 +332,23 @@ def transaction(self, *, savepoint: bool = False, interrupting: bool = False) ->
             yield
             trans.commit()
         except BaseException:
             trans.rollback()
```
If non-savepoint nested transaction is aborted then it aborts actual database transaction (meaning _all_ nested transactions are aborted too), so you also need to clear `self._cs.transactions`, not just `pop()` latest one.

andy-slac replied:
I think larger issue here is that even when you are in a savepoint transaction the underlying engine can decide to abort top-level transaction for any reason (e.g. Postgres aborts transactions for any error). In that case you somehow need to know what is the actual state of a backend transaction is and keep your `transactions` list consistent.

TallJimbo replied:
Hmm, good point.  And I think that means that non-nested transactions are just fundamentally incompatible with Python context managers, at least in terms of yielding non-surprising behavior.  I think I'm just going to remove the `savepoint` argument and just go back to always using `begin_nested` for now (as we do in `SqlRegistry`).  Hopefully just using bulk operations will still keep the number of actual SAVEPOINT commands low, but if not I think we need to look into some other way of expressing transactions in Python.

TallJimbo replied:
Hopefully switching to `Connection.in_transaction()` will take care of that.  If not, I think we'll have to see how the problem manifests to be able to do something about it.

andy-slac commented on python/lsst/daf/butler/registry/databases/oracle.py:
```diff
@@ -93,61 +93,55 @@ class OracleDatabase(Database):
+                             writeable: bool = True) -> Database:
         # TODO: is there anything we can do to make the connection read-only if
         # writeable is False?  If not (and at present) we just rely on the
         # Python code being well-behaved and not *trying* to make changes.
```
Oracle has `SET TRANSACTION READ ONLY` syntax (https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SET-TRANSACTION.html), I think this needs to be run as a first command in a transaction. It not only disables update commands but sets what is called "transaction-level read consistency" which means that all SELECTs view the data from the same snapshot (made at a time of transaction start).

 Postgres supports the same `SET TRANSACTION READ ONLY` (https://www.postgresql.org/docs/12/sql-set-transaction.html) but I could not find what exactly it does for that option.

TallJimbo replied:
I'll try to find SQLAlchemy event hooks where we can put those.

TallJimbo replied:
So, the event hooks exist, but there's no good way to get the information about whether the DB is read-only into them.  So instead I've just emitted that `SET TRANSACTION READ ONLY` statement at the beginning of transactions we start ourselves.  That by no means covers everything, but this is really just an extra guard against bad code, not the first line of defense.

andy-slac commented on python/lsst/daf/butler/registry/databases/postgresql.py:
```diff
@@ -55,26 +56,36 @@ class PostgresqlDatabase(Database):
+        if not writeable:
+            dbapi = cs.engine.raw_connection()
+            try:
+                dbapi.set_session(readonly=True)
```
I wonder if this works 100% reliably. I understand that connections are cached in a pool, but it may also happen that pool can decide to drop and re-establish a connection under some circumstances, and new connection will not have `readonly` option set. It may be better to pass a `creator` argument to `create_engine` or use `SET TRANSACTION READ ONLY` when transaction is started.

andy-slac commented on python/lsst/daf/butler/registry/databases/sqlite.py:
```diff
@@ -106,34 +107,133 @@ class SqliteDatabase(Database):
+            dbList = list(cursor.execute("PRAGMA database_list").fetchall())
+        if len(dbList) == 0:
+            raise RuntimeError("No database in connection.")
+        for _, dbname, filename in dbList:
```
I think you can simplify conditional logic below if you do `namespace = "main" if namespace is None else namespace`

andy-slac reviewed (approved):
Few minor (or not so minor) comments for this commit.

TallJimbo committed:
Simplify namespace lookup in SQLite.

TallJimbo committed:
Track transactions via Connection.in_transaction().

TallJimbo committed:
Automatically choose nested (savepoint) or non-nested transactions.

Nesting regular transactions doesn't actually work the way users would
expect, in that any rollback in an inner context also rolls back
the outer context.

TallJimbo committed:
Remove ConnectionStruct in favor of just using connections.

TallJimbo committed:
Attempt to make PostgreSQL and Oracle read-only safer.

This definitely doesn't cover all cases - only those where our own
transaction context manager is used - but it's still better than what
we had before.

TallJimbo reviewed (commented):
@andy-slac, up for one more round?  I'm now just using `Connection.in_transaction` (wish I'd realized that existed earlier), which I think addresses your biggest concern about as well as we can.  Combined with another change I discovered I needed to make on DM-22609 downstream of this, I realized I could just replace `ConnectionStruct` with directly using the SQLAlchemy `Connection` instance, since we don't need the engine either.  Six new small commits, some of which I'll squash later.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/database.py:
```diff
@@ -330,9 +326,11 @@ def transaction(self, *, savepoint: bool = False, interrupting: bool = False) ->
             "interrupt the active transaction context has been requested."
         )
-        if savepoint:
+        if self._cs.connection.in_transaction():
```
I guess this means we treat all nested transactions as savepoints. Savepoints have bigger overhead compared to non-savepoint "nested" sqlalchemy transactions, if we have a lot of nested transaction in registry this may cause performance issues.

TallJimbo replied:
Yup, understood.  As this is already a problem in the current `SqlRegistry`, I'm just punting the problem to a later ticket (I have some other mitigations in mind) as it's too tricky to solve quickly on this one.

andy-slac commented:
I checked recent commits, everything looks OK and waits to be battle-tested 🙂
